datalad_metalad/add.py

```diff
@@ -0,0 +1,449 @@
+        "dataset_id",
+        "dataset_version",
+        "extracted_metadata")
+
+    optional_keys = (
+        "intra_dataset_path",)
+
+    required_additional_keys = (
+        "root_dataset_id",
+        "root_dataset_version",
+        "inter_dataset_path")
+
+    required_keys_lines = "\n".join(map(repr, required_keys))
+    required_additional_keys_lines = "\n".join(
+        map(repr, required_additional_keys))
+
+    _params_ = dict(
+            metadata=Parameter(
+            args=("metadata",),
+            metavar="METADATA",
+            doc=f"""Path of a file that contains the metadata that
+            should be added to the metadata model instance (the
+            metadata must be provided as a JSON-serialized metadata
+            dictionary).
+            
+            If the path is "-", metadata is read from standard input.
+            
+            The dictionary must contain the following keys:
+            
+            {required_keys_lines}
+            
+            If the metadata is associated with a file, the following key
+            indicates the file path:
+            
+            'intra_dataset_path'
+            
+            It may in addition contain either all or none of the
+            following keys (they are used to add metadata element
+            as a sub-dataset element, i.e. perform aggregation):
+            
+            {required_additional_keys_lines}            
+            """,
+            constraints=EnsureStr() | EnsureNone()),
+        metadata_store=Parameter(
+            args=("-s", "--metadata-store"),
+            doc="""Directory in which the metadata model instance is
+            stored. If no directory name is provided, the current working
+            directory is used.""",
+            constraints=EnsureStr() | EnsureNone()),
+        additionalvalues=Parameter(
+            args=("additionalvalues",),
+            metavar="ADDITIONAL_VALUES",
+            doc="""A string that contains a JSON serialized dictionary of
+            key value-pairs. These key values-pairs are used in addition to
+            the key value pairs in the metadata dictionary to describe
+            the metadata that should be added. If an additional key is
+            already present in the metadata, it will override the value
+            from metadata. In this case a warning is issued.""",
+            nargs="?",
+            constraints=EnsureStr() | EnsureNone()),
+        allow_override=Parameter(
+            args=("-o", "--allow-override"),
+            doc="""Allow the additional values to override values given in
+            metadata.""",
+            default=False,
+            constraints=EnsureBool() | EnsureNone()),
+        allow_unknown=Parameter(
+            args=("-u", "--allow-unknown"),
+            doc="""Allow unknown keys. By default, unknown keys generate
+            an errors. If this switch is True, unknown keys will only be
+            reported.""",
```

Comment:
I assume this is for top-level keys only, and not for those in `extracted_metadata`?

Reply:
That is correct. Currently top-level keys only, and most likely not `extracted_metadata` (I think we have to see whether there would be a good use-case for changing metadata, maybe fixing errors introduced by different events?).

Reply:
refer to issue #96